File: test_c/calib3d_test.cpp
#include "gtest/gtest.h"
#include "opencv2/core.hpp"
using namespace cv;

#include "opencv_rice/calib3d/calib3d.h"
using namespace CVRice;

TEST( calculateHomography, AllPoints )
{
  const int rows = 10;
  Mat m1( rows, 2, CV_64F ), m2( rows, 2, CV_64F ), mask;

  for( int i = 0; i < rows; ++i ) {
    m1.at<double>(i,0) = sin(i);
    m1.at<double>(i,1) = cos(i);
    m2.at<double>(i,0) = sin(i);
    m2.at<double>(i,1) = cos(i)+1;
  }

  Mat h = CVRice::calculateHomography( m1, m2, mask, ALL_POINTS );
  ASSERT_EQ( 3, h.rows );
  ASSERT_EQ( 3, h.rows );

  Mat h_ransac = CVRice::calculateHomography( m1, m2, mask, CVRice::RANSAC );
  ASSERT_EQ( 3, h_ransac.rows );
  ASSERT_EQ( 3, h_ransac.rows );
  ASSERT_EQ( 10, mask.rows );
  ASSERT_EQ( 1, mask.cols );

  Mat h_lmeds = CVRice::calculateHomography( m1, m2, mask, CVRice::LMEDS );
  ASSERT_EQ( 3, h_lmeds.rows );
  ASSERT_EQ( 3, h_lmeds.rows );
  ASSERT_EQ( 10, mask.rows );
  ASSERT_EQ( 1, mask.cols );


//  ASSERT_EQ( CV_64F, err.type() );
//  const double exp = 17.0;
//  for( int j = 0; j < rows; ++j ) {
//    ASSERT_DOUBLE_EQ( exp, err.at<double>(j,0) );
//  }
}

File: test/calib3d_test.rb

require "opencv-rice"

require "minitest/autorun"

class TestCalib3d < Minitest::Test

    def setup
      @rows = 10
      # Must ensure points are non-colinear
      @m1 = Array.new(@rows) { |i| [ Math::sin(i), Math::cos(i) ] }
      @m2 = Array.new(@rows) { |i| [ Math::sin(i), Math::cos(i)+1 ] }
    end

    def test_find_homography
      h = CVRice::findHomography( @m1, @m2 )

      assert_equal 3, h.rows
      assert_equal 3, h.cols
      #rows.times.each do |i|
      #  assert_in_delta 45.0, err.at_d( i )
      #end
    end

    def test_find_homography_ransac
      h = CVRice::findHomography( @m1, @m2, CVRice::RANSAC )
      assert_equal 3, h.rows
      assert_equal 3, h.cols
    end

    def test_find_homography_lmeds
      h = CVRice::findHomography( @m1, @m2, CVRice::LMEDS )
      assert_equal 3, h.rows
      assert_equal 3, h.cols
    end
end



class TestPose < Minitest::Test
  include CVRice

  def pose_tests( pose )
    assert_instance_of CVRice::Pose, pose

    [ pose.rvec, pose.tvec ].each { |vec|
      assert_instance_of Mat, vec
      assert_equal 3, vec.rows
      assert_equal 1, vec.cols
    }

    rot = pose.rotation_matrix
    assert_instance_of Mat, rot
    assert_equal 3, rot.rows
    assert_equal 3, rot.cols

    total = pose.total
    assert_instance_of Mat, total
    assert_equal 4, total.rows
    assert_equal 4, total.cols
  end

  def test_initialise_rotation_matrix
    pose = Pose.new( Mat.identity(3), Mat.columns([ [0,0,0] ] ) )
    pose_tests( pose )
  end

  def test_initialise_rvec
    pose = Pose.new( Mat.columns([ [0,0,0] ]), Mat.columns([ [0,0,0] ] ) )
    pose_tests( pose )
  end
end


